triggers/service/authentication.md

```diff
@@ -0,0 +1,95 @@
+- Do not require OAuth2 or any other specific
+  authentication/authorization protocol from the IAM. In other words,
+  the communication with the IAM must be pluggable.
+- Do not rely on wildcard access for the trigger service, it should
+  only be able to start triggers on behalf of a party if a user that
+  controls that party has given consent.
+- Support long-running triggers without constant user
+  interaction. Since auth tokens are often short-lived (e.g., expire
+  after 1h), this implies some mechanism for token refresh.
+
+## Design
+
+This involves 3 components:
+
+1. The trigger service provided by DA.
+2. An auth middleware. DA provides an implementation of this for at
+   least the OAuth2 authorization code grant but this is completely
+   pluggable so if the DA-provided middleware does not cover the IAM
+   infrastructure of a client, they can implement their own.
+3. The IAM. This is the entity that signs Ledger API tokens. This is
+   not provided by DA. The Ledger is configured to trust this entity.
+
+### Auth Middleware API
+
+The auth middleware provides a few endpoints (the names don’t matter
+all that much, they just need to be fixed once).
+
+1. /auth The trigger service, will contact this endpoint with a set of
+   claims. If the user has already authenticade and is authorized for
+   those claims, it will return an access token (an opaque
+   blob to the trigger service) for at least those claims and a
+   refresh token (another opaque blob). If not, it will return an
+   unauthorized status code.
+
+2. /login If /auth returned unauthorized, the trigger service will
+   redirect users to this. The parameters will include the requested
+   claims as well as the URL that the users tried to access originally
+   (e.g., the endpoint to start a trigger). This will start an auth
+   flow, e.g., an OAuth2 authorization code grant ending with the auth
+   service setting a cookie with the access and refresh token and
+   redirecting to the original URL. At this point, the request to
+   /auth will succeed (based on the cookie).
+
+3. /refresh This accepts a refresh token and returns a new access
+   token and optionally a new refresh token (or fails).
+
+### Auth Middleware Implementation based on OAuth2 Authorization Code Grant
+
+1. /auth checks for the presence of a cookie with the tokens in it.
+2. /login starts an OAuth2 authorization code grant flow. After the
+   redirect URI is called by the authorization server, the middleware
+   makes a request to get the tokens, sets them in cookies and
+   redirects back to the original URI.
+3. /refresh simply proxies to the refresh endpoint on the
+   authorization server adding the client id and secret.
+
+Note that the auth middleware does not need to persist any state in
+this model. The trigger service does need to persist at least the
+refresh token and potentially the access token.
+
+## Issues
+
+1. This only works if the original request was a GET
```

Comment:
This is currently my biggest gripe with this approach. For the trigger service, GET requests are not completely unreasonable but it still feels very arbitrary and weird. I don’t have a great solution here so very happy to hear any suggestions.

Reply:
I think it's a reasonable restriction. This isn't a general-purpose middleware, why make things more complicated on the client side?

Reply:
i don't see a way to remove this restriction and still have this solution work in a browser seamlessly without removing the proxying behavior by replacing it with a trust relationship between an external website and the middleware component